### PR TITLE
Feat/merge distributed load

### PR DESCRIPTION
**Pull request by aNebula**
Implemented loading of distributed models and merge command.
## Types of changes
- \[ \] Bug fix (non-breaking change which fixes an issue)
- \[x\] New feature (non-breaking change which adds functionality)
- \[ \] Breaking change (fix or feature that would cause existing functionality to change)
- \[x\] My code follows the code style of this project.
- \[ \] My change requires a change to the documentation.
- \[x\] I have updated the documentation accordingly.
- \[x\] I have added tests to cover my changes.
- \[x\] All new and existing tests passed.
- \[x\] All commits are signed.

### CODE REVIEW

File: trestle/core/commands/add.py
Comment (butler54) [inline trestle/core/commands/add.py:73]: Can you please make sure an issue is created for this?

File: trestle/utils/fs.py
Comment (butler54) [inline trestle/utils/fs.py:164]: I don't know what the 'none' is doing here. I think given what I see below in like 173 we should just force people to pass it. Optional arguments for no good reason are a pain (and the type annotation needs to be wrapped in optional.

Comment (butler54) [reply trestle/utils/fs.py:164]: this is getting the model type. The output tuple was correct if it is passing a class and not a class instance.

File: trestle/core/commands/merge.py
Comment (butler54) [inline trestle/core/commands/merge.py:53]: can we please put a check in here rather than choosing to error out badly?

File: trestle/core/commands/merge.py
Comment (butler54) [inline trestle/core/commands/merge.py:75]: Do we think this is useful? Should it be added to a backlog?

File: trestle/core/commands/merge.py
Comment (butler54) [inline trestle/core/commands/merge.py:75]: This has assumed json. You should infer based on the other files which are being loaded.

Comment (butler54) [reply trestle/core/commands/merge.py:75]: or the file path provided with -f

File: trestle/core/commands/merge.py
Comment (butler54) [inline trestle/core/commands/merge.py:120]: Avoid using internals here (typing_extensions.get_origin()

File: trestle/core/commands/merge.py
Comment (butler54) [inline trestle/core/commands/merge.py:115]: You  have made many assumptions of json. at least parameterize this so it's easy to fix later.

File: trestle/core/commands/merge.py
Comment (butler54) [inline trestle/core/commands/merge.py:128]: Typing will be strange here.

File: tests/trestle/utils/load_distributed_test.py
Comment (butler54) [inline tests/trestle/utils/load_distributed_test.py:33]: Do we really need to do this? tmp_trestle_dir will fail if it is not created. We probably should kill that function.

File: tests/trestle/core/commands/merge_test.py
Comment (butler54) [inline tests/trestle/core/commands/merge_test.py:79]: you would be missing the . from .json

Comment (butler54) [reply tests/trestle/core/commands/merge_test.py:79]: i'm actually surprised this works.

**Review by butler54 — CHANGES_REQUESTED** [review]
Overall looks good. I think there are few small things that need to be fixed including typing. The extensions in testing and in merge look strange - So please double check.

File: tests/trestle/core/commands/merge_test.py
Comment (aNebula) [reply tests/trestle/core/commands/merge_test.py:79]: `fext = FileContentType.to_file_extension(content_type)` includes the .

File: tests/trestle/utils/load_distributed_test.py
Comment (aNebula) [reply tests/trestle/utils/load_distributed_test.py:33]: Need a initialised trestle directory to copy the testdata. Testdata does not include a initialised trestle project.

File: trestle/utils/load_distributed.py
Comment (aNebula) [reply trestle/utils/load_distributed.py:27]: > It would be great if you could provide a google docstring style 'args' and 'Returns' here as the return types are a bit vague from reading the code.
> 
> E.g. the OscalBaseModel is stripped.
Not stripped. Returned list of OscalBaseModel will be fully loaded with distributed elements.

Comment (aNebula) [reply trestle/utils/load_distributed.py:27]: > Actually on second thoughts maybe this should be an internal function rather than one intended to be exposed to the end user (aka leading underscore in the name).

Yeah. I thought so too at some point. Changed to internal function

File: trestle/utils/load_distributed.py
Comment (aNebula) [reply trestle/utils/load_distributed.py:60]: Going with Optional[Type[Any]]. Otherwise it's `Optional[Union[Type[List[OscalBaseModel]], Type[Dict[str, OscalBaseModel]]]]`, which looks horrible. Also I don't understand why mypy needs annotations for List items, when I'm using Type[List]. 😕

File: trestle/utils/load_distributed.py
Comment (aNebula) [reply trestle/utils/load_distributed.py:61]: Agreed and Done.

File: trestle/utils/load_distributed.py
Comment (aNebula) [reply trestle/utils/load_distributed.py:102]: No `get_origin()` defined on the typing_extensions 😕

File: trestle/utils/load_distributed.py
Comment (aNebula) [reply trestle/utils/load_distributed.py:110]: ✅

File: trestle/utils/fs.py
Comment (aNebula) [reply trestle/utils/fs.py:164]: I believe the None was by design - You can get contextual model without passing a path - in that case cwd is used.
The return types I agree - got taken out in git merge - added back.

File: trestle/core/commands/merge.py
Comment (aNebula) [reply trestle/core/commands/merge.py:53]: ✅

File: trestle/core/commands/merge.py
Comment (aNebula) [reply trestle/core/commands/merge.py:75]: Hmm... it could be useful. However, I did not investigate. May be we can keep the branch alive and bring back in future release?

File: trestle/core/commands/merge.py
Comment (aNebula) [reply trestle/core/commands/merge.py:75]: I have opened an issue for it. Let me handle them when I handle the issue.

File: trestle/core/commands/merge.py
Comment (aNebula) [reply trestle/core/commands/merge.py:120]: `get_origin()` Seems to be not defined 😕

File: trestle/core/commands/merge.py
Comment (aNebula) [reply trestle/core/commands/merge.py:115]: There's an issue for it - I'll handle it with the issue

File: trestle/core/commands/merge.py
Comment (aNebula) [reply trestle/core/commands/merge.py:128]: a `# type: ignore` case

File: trestle/utils/load_distributed.py
Comment (butler54) [reply trestle/utils/load_distributed.py:114]: Check line 85 of `trestle/core/generators.py`

**Review by butler54 — APPROVED** [review]
LGTM assuming the minor fixes which are outlined in the issues. 

Great job.

File: trestle/core/commands/merge.py
Comment (butler54) [reply trestle/core/commands/merge.py:122]: Created #278 to cover off this issue as there are some interdependencies which are messy.

File: trestle/utils/load_distributed.py
Comment (butler54) [reply trestle/utils/load_distributed.py:114]: Created #278 to cover off this issue as there are some interdependencies which are messy.

**Comment by butler54** [comment]
Final comment - I'm going to merge this now. Due to the codebase differences with Frank's branch it's going to be easier to deal with before we merge his OSCAL 1.0.0 release branch in.

**Pull request merged**
